portal/models/questionnaire_response.py

```diff
@@ -61,6 +63,65 @@ def __str__(self):
         return "QuestionnaireResponse {0.id} for user {0.subject_id} " \
                "{0.status} {0.authored}".format(self)
 
+    def assign_qb_relationship(self, acting_user_id):
+        """Lookup and assign questionnaire bank and iteration
+
+        On submission, and subsequently when a user's state changes (such as
+        the criteria for trigger_date), determine the associated questionnaire
+        bank and iteration and assign, or clear if no match is found.
+
+        :param acting_user_id: current driver of process, for audit purposes
+
+        """
+        from .qb_status import QB_Status  # avoid cycle
+        initial_qb_id = self.questionnaire_bank_id
+        initial_qb_iteration = self.qb_iteration
+
+        # clear both until current values are determined
+        self.questionnaire_bank_id, self.qb_iteration = None, None
+
+        authored = FHIR_datetime.parse(self.document['authored'])
+        qn_ref = self.document.get("questionnaire").get("reference")
+        qn_name = qn_ref.split("/")[-1] if qn_ref else None
+        qn = Questionnaire.find_by_name(name=qn_name)
+        qbstatus = QB_Status(self.subject, as_of_date=authored)
+        qbd = qbstatus.current_qbd()
+        if qbd and qn and qn.id in [
```

Comment:
I try to use tuples or generator comprehensions if I don't have any need for a list. Probably a premature optimization (little less memory), but I don't think there are many downsides.

Reply:
Okay, old habits...but a good suggestion.